File: scripts/PatratulLuiPitagora.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class PatratulLuiPitagora {
    private static final VectorDefinitie[] VECTORI = {
        new VectorDefinitie("Scara bunastarii materiale", new int[] {3, 6, 9}),
        new VectorDefinitie("Vectorul relational si social", new int[] {2, 5, 8}),
        new VectorDefinitie("Scara bunastarii spirituale", new int[] {1, 4, 7}),
        new VectorDefinitie("Axa personala", new int[] {3, 2, 1}),
        new VectorDefinitie("Axa de constructie", new int[] {6, 5, 4}),
        new VectorDefinitie("Axa superioara si sociala", new int[] {9, 8, 7}),
        new VectorDefinitie("Vectorul scopului", new int[] {3, 5, 7}),
        new VectorDefinitie("Vectorul carierei", new int[] {1, 5, 9})
    };

    public static void main(String[] args) {
        if (args.length != 3) {
            System.err.println("Utilizare: java PatratulLuiPitagora <zi> <luna> <an>");
            System.exit(1);
        }

        int zi = Integer.parseInt(args[0]);
        int luna = Integer.parseInt(args[1]);
        int an = Integer.parseInt(args[2]);

        Rezultat rezultat = calculeaza(zi, luna, an);
        afiseaza(rezultat);
    }

    private static Rezultat calculeaza(int zi, int luna, int an) {
        String data = String.format("%02d%02d%04d", zi, luna, an);
        List<Integer> cifreData = cifreDinText(data, true);

        int n1 = cifreData.stream().mapToInt(Integer::intValue).sum();
        int n2 = sumaCifrelor(n1);
        int n3 = n1 - 2 * primaCifraNenula(zi);
        int n4 = sumaCifrelor(n3);

        String sirComplet = data + " + " + n1 + " + " + n2 + " + " + n3 + " + " + n4;
        List<Integer> cifreMatrice = cifreDinText(data + n1 + n2 + n3 + n4, false);
        Map<Integer, String> matrice = grupeazaMatrice(cifreMatrice);
        List<VectorInterpretare> vectori = interpreteazaVectori(matrice);

        return new Rezultat(data, cifreData, n1, n2, n3, n4, sirComplet, cifreMatrice, matrice, vectori);
    }

    private static int sumaCifrelor(int numar) {
        return cifreDinText(String.valueOf(numar), true)
            .stream()
            .mapToInt(Integer::intValue)
            .sum();
    }

    private static int primaCifraNenula(int numar) {
        for (char cifra : String.format("%02d", numar).toCharArray()) {
            if (cifra != '0') {
                return Character.getNumericValue(cifra);
            }
        }
        throw new IllegalArgumentException("Ziua trebuie sa contina cel putin o cifra nenula.");
    }

    private static List<Integer> cifreDinText(String text, boolean includeZero) {
        List<Integer> cifre = new ArrayList<>();
        for (char caracter : text.toCharArray()) {
            if (Character.isDigit(caracter) && (includeZero || caracter != '0')) {
                cifre.add(Character.getNumericValue(caracter));
            }
        }
        return cifre;
    }

    private static Map<Integer, String> grupeazaMatrice(List<Integer> cifre) {
        Map<Integer, String> matrice = new HashMap<>();
        for (int cifra = 1; cifra <= 9; cifra++) {
            StringBuilder valoare = new StringBuilder();
            for (int cifraCurenta : cifre) {
                if (cifraCurenta == cifra) {
                    valoare.append(cifra);
                }
            }
            matrice.put(cifra, valoare.toString());
        }
        return matrice;
    }

    private static List<VectorInterpretare> interpreteazaVectori(Map<Integer, String> matrice) {
        List<VectorInterpretare> interpretari = new ArrayList<>();
        for (VectorDefinitie vector : VECTORI) {
            int pozitiiPrezente = 0;
            int aparitii = 0;

            for (int pozitie : vector.pozitii) {
                String valoare = matrice.get(pozitie);
                if (!valoare.isEmpty()) {
                    pozitiiPrezente++;
                    aparitii += valoare.length();
                }
            }

            String stare;
            if (pozitiiPrezente == 3) {
                stare = "activ";
            } else if (pozitiiPrezente == 2) {
                stare = "partial activ";
            } else if (pozitiiPrezente == 1) {
                stare = "slab";
            } else {
                stare = "latent";
            }

            List<String> analiza = new ArrayList<>();
            for (int pozitie : vector.pozitii) {
                analiza.add(pozitie + ":" + matrice.get(pozitie).length());
            }

            interpretari.add(new VectorInterpretare(
                vector.nume,
                vector.pozitii,
                pozitiiPrezente,
                aparitii,
                stare,
                analiza
            ));
        }
        interpretari.sort(
            Comparator.comparingInt((VectorInterpretare vector) -> vector.valoare)
                .thenComparingInt(vector -> vector.pozitiiPrezente)
                .thenComparing(vector -> vector.nume)
                .reversed()
        );
        return interpretari;
    }

    private static void afiseaza(Rezultat rezultat) {
        System.out.println("Cod extras din data nasterii:");
        System.out.println(rezultat.data);
        System.out.println();

        System.out.println("Cifre brute, fara zero:");
        System.out.println(join(faraZero(rezultat.cifreData)));
        System.out.println();

        System.out.println("Numarul 1 de lucru:");
        System.out.println(joinCuPlus(rezultat.cifreData) + " = " + rezultat.n1);
        System.out.println();

        System.out.println("Numarul 2 de lucru:");
        System.out.println(joinCuPlus(cifreDinText(String.valueOf(rezultat.n1), true)) + " = " + rezultat.n2);
        System.out.println();

        System.out.println("Numarul 3 de lucru:");
        int primaCifra = primaCifraNenula(Integer.parseInt(rezultat.data.substring(0, 2)));
        System.out.println(rezultat.n1 + " - (2 x " + primaCifra + ") = " + rezultat.n3);
        System.out.println();

        System.out.println("Numarul 4 de lucru:");
        System.out.println(joinCuPlus(cifreDinText(String.valueOf(rezultat.n3), true)) + " = " + rezultat.n4);
        System.out.println();

        System.out.println("Cod complet atasat:");
        System.out.println(rezultat.sirComplet);
        System.out.println();

        System.out.println("Cifre introduse in matrice, fara zero:");
        System.out.println(join(rezultat.cifreMatrice));
        System.out.println();

        System.out.println("Patratul lui Pitagora:");
        System.out.printf("%-4s | %-4s | %-4s%n", rezultat.matrice.get(3), rezultat.matrice.get(6), rezultat.matrice.get(9));
        System.out.printf("%-4s | %-4s | %-4s%n", rezultat.matrice.get(2), rezultat.matrice.get(5), rezultat.matrice.get(8));
        System.out.printf("%-4s | %-4s | %-4s%n", rezultat.matrice.get(1), rezultat.matrice.get(4), rezultat.matrice.get(7));
        System.out.println();

        System.out.println("Vectori:");
        for (int index = 0; index < rezultat.vectori.size(); index++) {
            VectorInterpretare vector = rezultat.vectori.get(index);
            System.out.println(
                (index + 1) + ". " + vector.nume + " (" + joinPozitii(vector.pozitii) + "): "
                    + vector.stare + ", " + vector.pozitiiPrezente + "/3 pozitii, "
                    + "valoare " + vector.valoare + " (" + String.join(", ", vector.analiza) + ")"
            );
        }

        if (!rezultat.vectori.isEmpty()) {
            VectorInterpretare vectorDominant = rezultat.vectori.get(0);
            VectorInterpretare vectorSlab = rezultat.vectori.get(rezultat.vectori.size() - 1);
            System.out.println();
            System.out.println("Analiza comuna a vectorilor:");
            System.out.println("Vector dominant: " + vectorDominant.nume + " cu valoarea " + vectorDominant.valoare + ".");
            System.out.println("Vector cel mai slab: " + vectorSlab.nume + " cu valoarea " + vectorSlab.valoare + ".");
        }
    }

    private static List<Integer> faraZero(List<Integer> cifre) {
        List<Integer> rezultat = new ArrayList<>();
        for (int cifra : cifre) {
            if (cifra != 0) {
                rezultat.add(cifra);
            }
        }
        return rezultat;
    }

    private static String join(List<Integer> cifre) {
        List<String> valori = new ArrayList<>();
        for (int cifra : cifre) {
            valori.add(String.valueOf(cifra));
        }
        return String.join(", ", valori);
    }

    private static String joinCuPlus(List<Integer> cifre) {
        List<String> valori = new ArrayList<>();
        for (int cifra : cifre) {
            valori.add(String.valueOf(cifra));
        }
        return String.join(" + ", valori);
    }

    private static String joinPozitii(int[] pozitii) {
        List<String> valori = new ArrayList<>();
        for (int pozitie : pozitii) {
            valori.add(String.valueOf(pozitie));
        }
        return String.join("-", valori);
    }

    private static class VectorDefinitie {
        String nume;
        int[] pozitii;

        VectorDefinitie(String nume, int[] pozitii) {
            this.nume = nume;
            this.pozitii = pozitii;
        }
    }

    private static class VectorInterpretare {
        String nume;
        int[] pozitii;
        int pozitiiPrezente;
        int valoare;
        String stare;
        List<String> analiza;

        VectorInterpretare(
            String nume,
            int[] pozitii,
            int pozitiiPrezente,
            int valoare,
            String stare,
            List<String> analiza
        ) {
            this.nume = nume;
            this.pozitii = pozitii;
            this.pozitiiPrezente = pozitiiPrezente;
            this.valoare = valoare;
            this.stare = stare;
            this.analiza = analiza;
        }
    }

    private static class Rezultat {
        String data;
        List<Integer> cifreData;
        int n1;
        int n2;
        int n3;
        int n4;
        String sirComplet;
        List<Integer> cifreMatrice;
        Map<Integer, String> matrice;
        List<VectorInterpretare> vectori;

        Rezultat(
            String data,
            List<Integer> cifreData,
            int n1,
            int n2,
            int n3,
            int n4,
            String sirComplet,
            List<Integer> cifreMatrice,
            Map<Integer, String> matrice,
            List<VectorInterpretare> vectori
        ) {
            this.data = data;
            this.cifreData = cifreData;
            this.n1 = n1;
            this.n2 = n2;
            this.n3 = n3;
            this.n4 = n4;
            this.sirComplet = sirComplet;
            this.cifreMatrice = cifreMatrice;
            this.matrice = matrice;
            this.vectori = vectori;
        }
    }
}

File: scripts/patratul_lui_pitagora.py
import argparse
import sys
import unicodedata


if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")


VECTORI = [
    ("Bunastare materiala", [3, 6, 9]),
    ("Bunastare sociala", [2, 5, 8]),
    ("Bunastare spirituala", [1, 4, 7]),
    ("Vectorul energetic", [1, 2, 3]),
    ("Vectorul vointa", [4, 5, 6]),
    ("Vectorul creativitatii", [7, 8, 9]),
    ("Vectorul scopului", [3, 5, 7]),
    ("Vectorul carierei", [1, 5, 9]),
]

ALFABET = {
    **dict.fromkeys("AJS", 1),
    **dict.fromkeys("BKT", 2),
    **dict.fromkeys("CLU", 3),
    **dict.fromkeys("DMV", 4),
    **dict.fromkeys("ENW", 5),
    **dict.fromkeys("FOX", 6),
    **dict.fromkeys("GPY", 7),
    **dict.fromkeys("HQZ", 8),
    **dict.fromkeys("IR", 9),
}


def suma_cifrelor(numar):
    numar = abs(numar)
    return sum(int(cifra) for cifra in str(numar))


def reducere(numar):
    curent = abs(numar)
    while curent > 9:
        curent = suma_cifrelor(curent)
    return curent


def prima_cifra_nenula(numar):
    for cifra in f"{numar:02d}":
        if cifra != "0":
            return int(cifra)
    raise ValueError("Ziua trebuie sa contina cel putin o cifra nenula.")


def grupeaza_matrice(cifre):
    return {cifra: str(cifra) * cifre.count(cifra) for cifra in range(1, 10)}


def normalizeaza_text(text):
    fara_diacritice = unicodedata.normalize("NFD", text)
    fara_diacritice = "".join(
        caracter for caracter in fara_diacritice
        if unicodedata.category(caracter) != "Mn"
    )
    return "".join(caracter for caracter in fara_diacritice.upper() if caracter.isalpha())


def valori_nume(nume):
    litere = normalizeaza_text(nume)
    return [(litera, ALFABET[litera]) for litera in litere if litera in ALFABET]


def componente_nume(nume):
    componente = []
    for parte in nume.replace("-", " ").split():
        litere = normalizeaza_text(parte)
        valori = [(litera, ALFABET[litera]) for litera in litere if litera in ALFABET]
        if valori:
            total = sum(valoare for _, valoare in valori)
            componente.append({
                "componenta": parte,
                "normalizat": "".join(litera for litera, _ in valori),
                "valori": valori,
                "total": total,
                "redus": reducere(total),
            })
    return componente


def interpreteaza_vectori(matrice):
    interpretari = []
    for vector in VECTORI:
        nume = vector[0]
        pozitii = vector[1]
        pozitii_prezente = [pozitie for pozitie in pozitii if matrice[pozitie]]
        valoare = sum(
            sum(int(cifra) for cifra in matrice[pozitie])
            for pozitie in pozitii
        )
        analiza = [
            f"{pozitie}:{sum(int(cifra) for cifra in matrice[pozitie]) if matrice[pozitie] else 0}"
            for pozitie in pozitii
        ]

        if len(pozitii_prezente) == 3:
            stare = "activ"
        elif len(pozitii_prezente) == 2:
            stare = "partial activ"
        elif len(pozitii_prezente) == 1:
            stare = "slab"
        else:
            stare = "latent"

        interpretari.append({
            "nume": nume,
            "pozitii": pozitii,
            "prezente": len(pozitii_prezente),
            "valoare": valoare,
            "stare": stare,
            "analiza": analiza,
        })
    return sorted(
        interpretari,
        key=lambda vector: (vector["valoare"], vector["prezente"], vector["nume"]),
        reverse=True,
    )


def calculeaza_patratul_lui_pitagora(zi, luna, an):
    data = f"{zi:02d}{luna:02d}{an:04d}"
    cifre_data = [int(cifra) for cifra in data]

    n1 = sum(cifre_data)
    n2 = suma_cifrelor(n1)
    n3 = abs(n1 - 2 * prima_cifra_nenula(zi))
    n4 = suma_cifrelor(n3)

    sir_complet = f"{data} + {n1} + {n2} + {n3} + {n4}"
    cifre_matrice = [
        int(cifra)
        for cifra in f"{data}{n1}{n2}{n3}{n4}"
        if cifra != "0"
    ]

    matrice = grupeaza_matrice(cifre_matrice)

    return {
        "data": data,
        "cifre_data": cifre_data,
        "cifre_data_fara_zero": [cifra for cifra in cifre_data if cifra != 0],
        "n1": n1,
        "n2": n2,
        "n3": n3,
        "n4": n4,
        "sir_complet": sir_complet,
        "cifre_matrice": cifre_matrice,
        "matrice": matrice,
        "vectori": interpreteaza_vectori(matrice),
    }


def calculeaza_matrice_nume(nume):
    componente = componente_nume(nume)
    valori = [valoare for componenta in componente for valoare in componenta["valori"]]
    cifre_litere = [valoare for _, valoare in valori]
    cifre_componente = [componenta["redus"] for componenta in componente]
    numar_exprimare = reducere(sum(cifre_componente))
    cifre_matrice = cifre_litere + [numar_exprimare]
    matrice = grupeaza_matrice(cifre_matrice)

    return {
        "nume": nume,
        "nume_normalizat": "".join(litera for litera, _ in valori),
        "valori": valori,
        "componente": componente,
        "total_litere": sum(cifre_litere),
        "total_componente": sum(cifre_componente),
        "numar_exprimare": numar_exprimare,
        "cifre_litere": cifre_litere,
        "cifre_componente": cifre_componente,
        "cifre_matrice": cifre_matrice,
        "matrice": matrice,
        "vectori": interpreteaza_vectori(matrice),
    }


def format_suma(cifre):
    return " + ".join(str(cifra) for cifra in cifre)


def afiseaza_rezultat(rezultat):
    matrice = rezultat["matrice"]

    print("Cod extras din data nasterii:")
    print(rezultat["data"])
    print()

    print("Cifre brute, fara zero:")
    print(", ".join(str(cifra) for cifra in rezultat["cifre_data_fara_zero"]))
    print()

    print("Numarul 1 de lucru:")
    print(f"{format_suma(rezultat['cifre_data'])} = {rezultat['n1']}")
    print()

    print("Numarul 2 de lucru:")
    print(f"{format_suma([int(cifra) for cifra in str(rezultat['n1'])])} = {rezultat['n2']}")
    print()

    print("Numarul 3 de lucru:")
    prima_cifra = prima_cifra_nenula(int(rezultat["data"][:2]))
    print(f"{rezultat['n1']} - (2 x {prima_cifra}) = {rezultat['n3']}")
    print()

    print("Numarul 4 de lucru:")
    print(f"{format_suma([int(cifra) for cifra in str(rezultat['n3'])])} = {rezultat['n4']}")
    print()

    print("Cod complet atasat:")
    print(rezultat["sir_complet"])
    print()

    print("Cifre introduse in matrice, fara zero:")
    print(", ".join(str(cifra) for cifra in rezultat["cifre_matrice"]))
    print()

    print("Patratul lui Pitagora:")
    print(f"{matrice[1]:<4} | {matrice[4]:<4} | {matrice[7]:<4}")
    print(f"{matrice[2]:<4} | {matrice[5]:<4} | {matrice[8]:<4}")
    print(f"{matrice[3]:<4} | {matrice[6]:<4} | {matrice[9]:<4}")
    print()

    print("Vectori:")
    for index, vector in enumerate(rezultat["vectori"], start=1):
        pozitii = "-".join(str(pozitie) for pozitie in vector["pozitii"])
        print(
            f"{index}. {vector['nume']} ({pozitii}): "
            f"{vector['stare']}, {vector['prezente']}/3 pozitii, "
            f"valoare {vector['valoare']} ({', '.join(vector['analiza'])})"
        )

    if rezultat["vectori"]:
        vector_dominant = rezultat["vectori"][0]
        vector_slab = rezultat["vectori"][-1]
        print()
        print("Analiza comuna a vectorilor:")
        print(
            f"Vector dominant: {vector_dominant['nume']} "
            f"cu valoarea {vector_dominant['valoare']}."
        )
        print(
            f"Vector cel mai slab: {vector_slab['nume']} "
            f"cu valoarea {vector_slab['valoare']}."
        )


def afiseaza_matrice(titlu, matrice):
    print(titlu)
    print(f"{matrice[1]:<4} | {matrice[4]:<4} | {matrice[7]:<4}")
    print(f"{matrice[2]:<4} | {matrice[5]:<4} | {matrice[8]:<4}")
    print(f"{matrice[3]:<4} | {matrice[6]:<4} | {matrice[9]:<4}")
    print()


def afiseaza_rezultat_nume(rezultat):
    print()
    print("Matricea numelui:")
    print(f"Nume: {rezultat['nume']}")
    print(f"Nume normalizat: {rezultat['nume_normalizat']}")
    print()

    print("Valori litere:")
    print(", ".join(f"{litera}={valoare}" for litera, valoare in rezultat["valori"]))
    print()

    print("Componente reduse:")
    for componenta in rezultat["componente"]:
        print(
            f"{componenta['normalizat']}: "
            f"{format_suma([valoare for _, valoare in componenta['valori']])} "
            f"= {componenta['total']} -> {componenta['redus']}"
        )
    print()

    print("Numar de exprimare:")
    print(
        f"{format_suma(rezultat['cifre_componente'])} = "
        f"{rezultat['total_componente']} -> {rezultat['numar_exprimare']}"
    )
    print()

    print("Cifre introduse in matricea numelui:")
    print(", ".join(str(cifra) for cifra in rezultat["cifre_matrice"]))
    print()

    afiseaza_matrice("Patratul numelui:", rezultat["matrice"])

    print("Vectori nume:")
    for index, vector in enumerate(rezultat["vectori"], start=1):
        pozitii = "-".join(str(pozitie) for pozitie in vector["pozitii"])
        print(
            f"{index}. {vector['nume']} ({pozitii}): "
            f"{vector['stare']}, {vector['prezente']}/3 pozitii, "
            f"valoare {vector['valoare']} ({', '.join(vector['analiza'])})"
        )


def main():
    parser = argparse.ArgumentParser(
        description="Calculeaza codul si patratul lui Pitagora dupa metoda proiectului."
    )
    parser.add_argument("zi", type=int)
    parser.add_argument("luna", type=int)
    parser.add_argument("an", type=int)
    parser.add_argument("--nume")
    args = parser.parse_args()

    rezultat = calculeaza_patratul_lui_pitagora(args.zi, args.luna, args.an)
    afiseaza_rezultat(rezultat)
    if args.nume:
        rezultat_nume = calculeaza_matrice_nume(args.nume)
        afiseaza_rezultat_nume(rezultat_nume)


if __name__ == "__main__":
    main()
